### PR TITLE
Create CI to build & push PyBaMM's Docker Images

### PR DESCRIPTION
**Pull request by arjxn-py**
# Description

This PR aims to create a CI that builds all PyBaMM's docker images & then pushes them to DockerHub.

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by arjxn-py** [comment]
Related to #3312 
cc: @agriyakhetarpal @brosaplanella @Saransh-cpp

**Comment by arjxn-py** [comment]
I'd also need help with setting up the docker secrets for this one.

**Review by agriyakhetarpal — COMMENTED** [review]
Would building and pushing images in their separate jobs be a better way, rather than in a series of steps?

File: .github/workflows/docker.yml
Comment (agriyakhetarpal) [inline .github/workflows/docker.yml:7]: This should point to the `develop` branch.

File: .github/workflows/docker.yml
Comment (agriyakhetarpal) [inline .github/workflows/docker.yml:34]: We would need to use the official Docker action to push images (`docker/build-push-action`) and point it to a path where the built images eill be stored

**Comment by arjxn-py** [comment]
> Would building and pushing images in their separate jobs be a better way, rather than in a series of steps?

Can they run parallelly in case of separate jobs?

**Comment by arjxn-py** [comment]
> > Would building and pushing images in their separate jobs be a better way, rather than in a series of steps?
> 
Can they run parallelly in case of separate jobs? I'd be happy to switch to separate jobs either way though 🙂 
Thanks!

File: .github/workflows/docker.yml
Comment (arjxn-py) [reply .github/workflows/docker.yml:34]: Thanks, will look into it.

**Comment by agriyakhetarpal** [comment]
> Can they run parallelly in case of separate jobs?

Yes, they will run and upload images in parallel. We would need to check out the repository in each job, however. Since it would be developers who would be using the image, could you remove the sparse checkout and let it fetch the entire commit depth of the repository?

**Comment by arjxn-py** [comment]
> could you remove the sparse checkout and let it fetch the entire commit depth of the repository?

Yes, sure.

**Comment by agriyakhetarpal** [comment]
Actually, multiple jobs might not be needed in this case, a matrix for the solver arguments should suffice after the check out step

**Comment by Saransh-cpp** [comment]
> I'd also need help with setting up the docker secrets for this one.

Do you want us to add them in the GitHub repo or do you want us to generate it from Dockerhub?

**Comment by arjxn-py** [comment]
> Do you want us to add them in the GitHub repo or do you want us to generate it from Dockerhub?

I'd need help with adding them to github repo, i can generate & send secrets to you on slack.

File: .github/workflows/docker.yml
Comment (Saransh-cpp) [inline .github/workflows/docker.yml:22]: Added these secrets in the PyBaMM repo. Let me know if everything works.

Comment (arjxn-py) [reply .github/workflows/docker.yml:21]: This CI is working on my fork [here](https://github.com/arjxn-py/PyBaMM/actions/runs/6266390975/job/17017264690?pr=6).
Can you please once confirm if secrets are set how we want them to?

Comment (Saransh-cpp) [reply .github/workflows/docker.yml:21]: Yes! We'll only be able to test it once this PR is merged. The secrets are not passed down to PRs from forks for security reasons.

File: .github/workflows/docker.yml
Comment (Saransh-cpp) [inline .github/workflows/docker.yml:7]: We should not run this on PRs as secrets are not passed down to PRs for security reasons. This should only run on a push to develop branch.

**Comment by agriyakhetarpal** [comment]
Note: to push to Docker Hub we should use https://github.com/docker/build-push-action.

I noticed that they have some examples about setting up QEMU, we should be able to build both `AMD64` and `ARM64` with that (I may take a look in another PR).

File: .github/workflows/docker.yml
Comment (arjxn-py) [reply .github/workflows/docker.yml:7]: > Yes I was just running it on this PR in order to test, this workflow can be triggered manually or run periodically afterwards, whichever preferred.

Comment (Saransh-cpp) [reply .github/workflows/docker.yml:7]: Should be ready to merge once `pull_request` is removed. Now that I think of it, we should also add a cron job that builds and pushes docker images nightly? Pushing on every commit might be a bit excessive.

**Review by Saransh-cpp — APPROVED** [review]
Thanks, @arjxn-py! Looks great! Should be ready to merge once the comment below is resolved -

File: .github/workflows/docker.yml
Comment (agriyakhetarpal) [reply .github/workflows/docker.yml:7]: I think we should also check out other repositories and projects as to how they do it (most related workflows push to the development branch and overwrite the previous image). Upon multiple pushes to `develop` we can cancel intermediate builds here as we do for the unit tests and benchmarks workflows

Comment (Saransh-cpp) [reply .github/workflows/docker.yml:7]: > I think we should also check out other repositories and projects as to how they do it

+1!

Comment (agriyakhetarpal) [reply .github/workflows/docker.yml:7]: The first workflow I looked at was `PyHF`'s, of course. They build docker images on PRs to their development branch (`main` in their case), pushes to `main`, scheduled nightly, on version tags, for releases, and also have a workflow dispatch trigger.

Building and pushing images on release tags _might_ require configuring the naming scheme of the images to include the version since we would not want to overwrite any existing images. We should look into that before `23.9` comes out (additionally `23.9rcX`). For now, I think just the two options: pushes to `develop` and a workflow dispatch trigger should be fine, with concurrency checks to cancel the build in case two PRs are merged one after the other. The primary disadvantage of the nightlies is that it could cause some hinderances for developers working within their containers in case a large PR or one with breaking changes gets merged and the Dockerfile is not updated, though this would be a rare occurrence—developers have to deal with merge conflicts sometimes anyway, and this is the case without Docker as well.

It is possible to exercise more fine-grained control over how we can push to Docker Hub. Some data science projects use a workflow call trigger with particular event webhooks to run a workflow from other workflows. To say it is complicated is an understatement but this way, in theory, one can build the Docker images only at the completion of a particular workflow by calling another workflow through it. In this case, it would be when the tests pass on merging a PR to `develop`, which, in-turn, will be cancelled on subsequent pushes due to the concurrency setting (or do we not run CI on merge commits anymore?) We should explore this option only if the solution above seems to be excessive or just switch to the nightlies.

Comment (Saransh-cpp) [reply .github/workflows/docker.yml:7]: `push` and `workflow_dispatch` sounds good to me as well!

**Comment by arjxn-py** [comment]
> storage limits on Docker Hub

As of my knowledge & research there's no [storage limit for images on DockerHub](https://forums.docker.com/t/does-docker-hub-have-a-size-limitation-on-repos-or-images/10154) (i.e. we can have as many tags we want) however there's an individual image size limit which we don't have to bother about as our image size is smaller.

**Comment by agriyakhetarpal** [comment]
That's great, thanks! Pushing all images should be fine then

**Review by Saransh-cpp — APPROVED** [review]

**Review by agriyakhetarpal — APPROVED** [review]

File: .github/workflows/docker.yml
Comment (agriyakhetarpal) [inline .github/workflows/docker.yml:72]: I think we did not tag the images properly. On https://hub.docker.com/r/pybamm/pybamm/tags I can see that this image was pushed with the `latest` tag, and

File: .github/workflows/docker.yml
Comment (agriyakhetarpal) [inline .github/workflows/docker.yml:29]: this one **also** has the `latest` tag

Comment (agriyakhetarpal) [reply .github/workflows/docker.yml:29]: the earlier one (with all solvers) should have had the `all` tag. I do have a local branch which fixes things up a bit for this workflow (including logging in and uploading these images in parallel jobs) and will write a PR soon, I might look into adding the multi-platform images there itself.

**Pull request merged**
